presto-spi/src/main/java/com/facebook/presto/spi/block/PageEncodingBuffers.java

```diff
@@ -0,0 +1,140 @@
+import static java.lang.Math.max;
+
+public class PageEncodingBuffers
+{
+    private int channelCount;
+    int bufferedRowCount;
+    BlockEncodingBuffers[] blockEncodingBuffers;
+
+    private int capacity;
+
+    public PageEncodingBuffers()
+    {
+        channelCount = 0;
+        bufferedRowCount = 0;
+        blockEncodingBuffers = null;
+        capacity = 0;
+    }
+
+    public boolean isEmpty()
+    {
+        return bufferedRowCount == 0;
+    }
+
+    public boolean isFull()
+    {
+        return capacity <= getSizeInBytes();
+    }
+
+    public void prepareBuffers(Page page)
+    {
+        capacity = DEFAULT_MAX_PAGE_SIZE_IN_BYTES;
+
+        if (blockEncodingBuffers == null || this.channelCount != page.getChannelCount()) {
+            channelCount = page.getChannelCount();
+            blockEncodingBuffers = new BlockEncodingBuffers[channelCount];
+            for (int i = 0; i < channelCount; i++) {
+                blockEncodingBuffers[i] = createBlockEncodingBuffers(page.getBlock(i));
+            }
+        }
+
+        for (int i = 0; i < channelCount; i++) {
+            blockEncodingBuffers[i].prepareBuffers();
+        }
+    }
+
+    public void appendRows(Page page, int[] positions, int startPositionsIndex, int batchSize)
+    {
+        for (int i = 0; i < page.getChannelCount(); i++) {
+            Block block = page.getBlock(i);
+            blockEncodingBuffers[i].copyPositions(positions, startPositionsIndex, batchSize);
+            block.appendValuesToBuffers(blockEncodingBuffers[i]);
+            blockEncodingBuffers[i].bufferedRowCount += batchSize;
+        }
+        bufferedRowCount += batchSize;
+    }
+
+    public BlockEncodingBuffers getBlockEncodingBuffers(int position)
+    {
+        // TODO: validate position
+        return blockEncodingBuffers[position];
+    }
+
+    public int calculateNumberRowsFitInBuffers(int[] rowSizes, int numberOfRows, int firstRowIndex)
+    {
+        // TODO: validate rows and rowSizes length >= numberOfRows > firstRowIndex
+        int roomInBytes = capacity - getSizeInBytes();
+
+        int rowIndex = firstRowIndex;
+        int totalSize = 0;
+        while (rowIndex < numberOfRows && totalSize <= roomInBytes) {
+            totalSize += rowSizes[rowIndex++];
+        }
+        int rowsFit = totalSize <= roomInBytes ? rowIndex - firstRowIndex : rowIndex - firstRowIndex - 1;
+        return max(rowsFit, 1);
+    }
+
+    // Copy the buffers in order into the final slice
+    public Slice serialize()
+    {
+        // TODO: Validate the bufferedRowCount > 0
+        //  The performace using DynamicSliceOutput should be ok if the capacity is enough, and for each buffer we only do one range check.
+        // TODO: size should include channelCount etc
+        int sizeInBytes = getSizeInBytes();
+        byte[] finalSerializedBytes = new byte[sizeInBytes];
+        Slice slice = Slices.wrappedBuffer(finalSerializedBytes);
```

Comment:
This can be simplified to `Slice slice = Slices.wrappedBuffer(new byte[getSizeInBytes()]);`, but why not use DynamicSliceOutput?

Reply:
DynamicSliceOutput.writeXXX() call Slices.ensureSize() every time. But for us we know the sizes in advance and don't need to do it. It's mainly a performance consideration

Reply:
@yingsu00 Can this be solved by adding batch APIs to DynamicSliceOutput?